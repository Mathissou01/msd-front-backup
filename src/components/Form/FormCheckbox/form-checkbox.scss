@use "src/styles/01-settings/settings.colors" as colors;

.c-FormCheckbox {
  display: flex;
  flex-direction: row;
  gap: 8px;
  align-items: center;
  padding: 4px;

  &__Input {
    box-sizing: border-box;
    width: 24px;
    height: 24px;
    border: 2px solid colors.$expert-blue;
    border-radius: 4px;
    cursor: pointer;
    appearance: none;

    &_checked {
      background: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' width='8' height='8' viewBox='0 0 8 8'%3e%3cpath fill='%23030f40' d='M6.564.75l-3.59 3.612-1.538-1.55L0 4.26l2.974 2.99L8 2.193z'/%3e%3c/svg%3e")
        no-repeat;
      background-position: 50% 50%;
      background-size: 60%;
    }
  }

  &__Label {
    color: colors.$expert-blue;
    font-weight: 400;
    font-size: 16px;
    cursor: pointer;
  }
}
