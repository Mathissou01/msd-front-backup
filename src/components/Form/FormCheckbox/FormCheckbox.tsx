import { useFormContext } from "react-hook-form";
import { ErrorMessage } from "@hookform/error-message";
import React from "react";
import FormLabel from "../FormLabel/FormLabel";
import "./form-checkbox.scss";

interface IFormCheckboxProps {
  name: string;
  label: string;
  secondaryLabel?: string;
  defaultChecked?: boolean;
  isChecked?: boolean;
  isRequired?: boolean;
  isDisabled?: boolean;
  onClick?: () => void;
}

export default function FormCheckbox({
  name,
  label,
  secondaryLabel,
  defaultChecked,
  isDisabled,
  isRequired,
  isChecked,
  onClick,
}: IFormCheckboxProps) {
  /* Local Data */
  const {
    register,
    formState: { isSubmitting, errors },
    watch,
  } = useFormContext();

  const watchChecked = watch(name, isRequired ? isChecked : defaultChecked);

  return (
    <div className="c-FormCheckbox">
      <input
        className={`c-FormCheckbox__Input ${
          watchChecked ? "c-FormCheckbox__Input_checked" : ""
        }`}
        type="checkbox"
        {...register(name)}
        id={name}
        defaultChecked={defaultChecked}
        disabled={isSubmitting || isDisabled}
        data-testid="form-checkbox"
        onClick={onClick}
      />
      <FormLabel forId={name} label={label} secondaryLabel={secondaryLabel} />
      <ErrorMessage
        errors={errors}
        name={name}
        render={({ message }: { message: string }) => (
          <p>{message ?? "Error"}</p>
        )}
      />
    </div>
  );
}
