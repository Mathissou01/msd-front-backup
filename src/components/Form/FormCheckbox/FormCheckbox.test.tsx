import { fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import React, { ReactNode } from "react";
import FormCheckbox from "./FormCheckbox";

const mock = {
  name: "form-checkbox",
  label: "form checkbox",
  secondaryLabel: "secondary label",
};

const Wrapper = (props: { children: ReactNode }) => {
  const formMethods = useForm();
  return <FormProvider {...formMethods}>{props.children},</FormProvider>;
};

describe("FormCheckbox", () => {
  it("renders", () => {
    const { container } = render(
      <Wrapper>
        <FormCheckbox
          name={mock.name}
          label={mock.label}
          secondaryLabel={mock.secondaryLabel}
          defaultChecked={false}
          isDisabled={false}
        />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-checkbox");
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).not.toHaveClass("c-FormCheckbox__Input_checked");
    expect(checkbox).not.toHaveAttribute("disabled");

    expect(container).toMatchSnapshot();
  });

  it("changes style when checked", async () => {
    render(
      <Wrapper>
        <FormCheckbox
          name={mock.name}
          label={mock.label}
          defaultChecked={true}
        />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-checkbox");
    expect(checkbox).toHaveClass("c-FormCheckbox__Input_checked");
    fireEvent.click(checkbox);
    expect(checkbox).not.toHaveClass("c-FormCheckbox__Input_checked");
  });

  it("can be disabled", () => {
    render(
      <Wrapper>
        <FormCheckbox name={mock.name} label={mock.label} isDisabled />
      </Wrapper>,
    );

    const checkbox = screen.getByTestId("form-checkbox");
    expect(checkbox).toHaveAttribute("disabled");
    fireEvent.click(checkbox);
    expect(checkbox).not.toHaveClass("c-FormCheckbox__Input_checked");
  });
});
