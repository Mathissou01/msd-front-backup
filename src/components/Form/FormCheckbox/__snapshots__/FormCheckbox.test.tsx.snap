// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckbox renders 1`] = `
<div>
  <div
    class="c-FormCheckbox"
  >
    <input
      class="c-FormCheckbox__Input "
      data-testid="form-checkbox"
      id="form-checkbox"
      name="form-checkbox"
      type="checkbox"
    />
    <div
      class="c-FormLabel"
    >
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-checkbox"
      >
        <span
          class="c-FormLabel__Label"
        >
          form checkbox
        </span>
      </label>
      <label
        class="c-FormLabel__LabelWrapper"
        for="form-checkbox"
      >
        <span
          class="FormLabel__Secondary"
        >
          secondary label
        </span>
      </label>
    </div>
  </div>
  ,
</div>
`;
